### PR TITLE
Encoder special symbols

### PR DESCRIPTION
**Pull request by jindrahelcl**
This PR changes the behavior of the sentence encoder - now, it does not attach `<s>` and `</s>` to the sentence. This is the behavior of both nematus and the tensorflow nmt model. 

Moreover, the encoder was refactored, documented and type-annotated. The `dropout_keep_p` vs `dropout_keep_prob` was once and for all united on `dropout_keep_prob` in both decoder and encoders

Dropout is removed from `decoding_function` and computed from the encoder (factored encoder was also changed)

In future, input tensors could be changed to time major shape mode so there are no unnecessary transpositions.

### CODE REVIEW

File: neuralmonkey/decoders/monster_decoder.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 class Decoder(object):
```
Comment (jlibovicky) [inline neuralmonkey/decoders/monster_decoder.py]: Kde se tu zase vzalo tohle monstrum? Už jsem ho jednou smazal :-D

File: neuralmonkey/encoders/sentence_encoder.py [hunk 2]
```diff
@@ -1,139 +1,211 @@
...
 from neuralmonkey.nn.noisy_gru_cell import NoisyGRUCell
-from neuralmonkey.nn.pervasive_dropout_wrapper import PervasiveDropoutWrapper
 from neuralmonkey.nn.ortho_gru_cell import OrthoGRUCell
-from neuralmonkey.checking import assert_type
+from neuralmonkey.dataset import Dataset
 from neuralmonkey.vocabulary import Vocabulary
 
+# pylint: disable=invalid-name
+AttType = Any ## Type[] or union of types do not work here
+FeedDict = Dict[tf.Tensor, Any]
```
Comment (jlibovicky) [inline neuralmonkey/encoders/sentence_encoder.py]: FeedDict bude takhle už někde zavedenej, tak bych ho jenom importoval.

File: neuralmonkey/encoders/sentence_encoder.py [hunk 3]
```diff
@@ -1,139 +1,211 @@
...
+            fw_cell, bw_cell = self.rnn_cells() # type: RNNCellTuple
 
-            if parent_encoder:
-                self.word_embeddings = parent_encoder.word_embeddings
-            else:
-                # NOTE the same note as in decoder when initializing embeddings
-                self.word_embeddings = tf.get_variable(
-                    "word_embeddings", shape=[len(vocabulary), embedding_size],
-                    initializer=tf.random_normal_initializer(stddev=0.01))
+            outputs_bidi_tup, encoded_tup = tf.nn.bidirectional_dynamic_rnn(
```
Comment (jlibovicky) [inline neuralmonkey/encoders/sentence_encoder.py]: Cool.

File: neuralmonkey/vocabulary.py [hunk 4]
```diff
@@ -14,6 +14,10 @@
 from neuralmonkey.logging import log
 from neuralmonkey.dataset import Dataset, LazyDataset
 
+#pylint: disable=invalid-name
+Sentence = List[str]
+#pylint: enable=invalid-name
```
Comment (jlibovicky) [inline neuralmonkey/vocabulary.py]: This one is unnecessary.

File: neuralmonkey/encoders/sentence_encoder.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: Já přemejšlel vůbec udělat modul `neuralmonkey.typing`, kde to bude všechno a odkud se to bude všechno importovat

File: neuralmonkey/decoders/monster_decoder.py (hunk 1 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/monster_decoder.py]: To je divný, já ho tam měl pořád. :-) Nechal bych ho tam dokud nereplikujem výsledky, kterých se s nim dosáhlo.. :-) Ale je fakt, že ta nejdůležitější část je ta decoding_function, kterou už mam vyrefaktorovanou v jiný branchi...

File: neuralmonkey/vocabulary.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/vocabulary.py]: ok
